### PR TITLE
Core: Fix querying metadata tables with multiple specs

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds a utility method to derive a common type for all partition specs.

Prior to this change, querying metadata tables in v2 format with evolved partitioning failed with runtime exceptions.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
cc @RussellSpitzer @rdblue @openinx @kbendick @jackye1995 @yyanyy @flyrain @karuppayya

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        .build();
+    TestTables.TestTable table = TestTables.create(tableDir, "test", SCHEMA, initialSpec, 1);
+
+    PartitionSpec newSpec = PartitionSpec.builderFor(table.schema())
+        .identity("category")
+        .build();
+
+    TableOperations ops = ((HasTableOperations) table).operations();
+    TableMetadata current = ops.current();
+    ops.commit(current, current.updatePartitionSpec(newSpec));
```
Comment (flyrain) [inline core/src/test/java/org/apache/iceberg/TestPartitioning.java:70]: Can we make change for the method `updatePartitionSpec` to avoid conflicts?

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestMetadataTablesWithPartitionEvolution.java [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    for (MetadataTableType tableType : Arrays.asList(ENTRIES, ALL_ENTRIES)) {
+      assertPartitionType(
+          ImmutableList.of(row(null, null), row("b1", null), row("b1", 2)),
+          "STRUCT<data:STRING,category_bucket_8:INT>",
+          tableType);
+    }
+
+    table.updateSpec()
+        .removeField("data")
+        .commit();
```
Comment (flyrain) [inline spark3/src/test/java/org/apache/iceberg/spark/source/TestMetadataTablesWithPartitionEvolution.java:219]: Nit: could be in one line.

**Review by flyrain — APPROVED** [review]
LGTM

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java (hunk 1 above)
Comment (aokolnychyi) [reply core/src/test/java/org/apache/iceberg/TestPartitioning.java:70]: This API is hidden from users. The user-facing API is `UpdatePartitionSpec` accessible via `Table`. That one actually ensures we don't hit this case. The spec evolution in v1 tables is actually limited as described [here](https://iceberg.apache.org/spec/#partitioning).

There could be some tables where people evolved partitioning before the public API appeared. It is an edge case but this test ensures we get a reasonable exception for such tables.

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 3]
```diff
@@ -177,4 +184,35 @@ public Void alwaysNull(int fieldId, String sourceName, int sourceId) {
       return null;
     }
   }
+
+  /**
+   * Builds a common partition type for all specs in a table.
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:189]: Could we add some detail in here? Seems like we are building a Struct which contains every field which has been part of a partitionspec in the past. So I would probably just write "Builds a struct containing all columns which have been ever been a part of a partition spec in this table"

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 4]
```diff
@@ -177,4 +184,35 @@ public Void alwaysNull(int fieldId, String sourceName, int sourceId) {
...
+    for (PartitionSpec spec : table.specs().values()) {
+      List<NestedField> fields = spec.partitionType().fields();
+
+      for (NestedField field : fields) {
+        int fieldId = field.fieldId();
+        NestedField existingField = commonFields.get(fieldId);
+
+        // partition fields may conflict in v1 tables
+        ValidationException.check(existingField == null || existingField.equals(field),
+            "Specs contain conflicting partition fields with ID %d",
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:206]: Could output the existing field here as well to illustrate the conflict?

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  public void setupTableDir() throws IOException {
+    this.tableDir = temp.newFolder();
+  }
+
+  @After
+  public void cleanupTables() {
+    TestTables.clearTables();
+  }
+
+  @Test
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/TestPartitioning.java:57]: Should we add a "works as expected" test here as well? I know we have it covered in the test metadata tables suite below but since we made this new file I thought might as well?

**Comment by RussellSpitzer** [comment]
Do we have to worry about the Manifests Table? Or is it ok because we are always displaying in the context of the current spec?

**Review by RussellSpitzer — APPROVED** [review]
A few minor comments, but looks good to me overall

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 4 above)
Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:206]: +1, can just use `"... with %s", field`

**Review by jackye1995 — APPROVED** [review]
looks good to me, thanks for the fix!

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestMetadataTablesWithPartitionEvolution.java [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    // verify the metadata tables after dropping the first partition column
+    for (MetadataTableType tableType : Arrays.asList(ENTRIES, ALL_ENTRIES)) {
+      assertPartitionType(
+          ImmutableList.of(row(null, null), row(null, 2), row("b1", null), row("b1", 2)),
+          "STRUCT<data:STRING,category_bucket_8:INT>",
+          tableType);
+    }
+  }
+
+  private void assertPartitionType(List<Object[]> expectedPartitions, String expectedTypeAsString,
```
Comment (karuppayya) [inline spark3/src/test/java/org/apache/iceberg/spark/source/TestMetadataTablesWithPartitionEvolution.java:232]: nit: assertPartitionType -> assertPartitionTypeAndValue

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java (hunk 1 above)
Comment (flyrain) [reply core/src/test/java/org/apache/iceberg/TestPartitioning.java:70]: I see, this part, https://iceberg.apache.org/spec/#partition-evolution.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 3 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java:189]: Added one more sentence below to clarify.

File: core/src/main/java/org/apache/iceberg/Partitioning.java (hunk 4 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/Partitioning.java]: Fixed.

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java (hunk 5 above)
Comment (aokolnychyi) [reply core/src/test/java/org/apache/iceberg/TestPartitioning.java:62]: Added two more tests to validate the column IDs.

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestMetadataTablesWithPartitionEvolution.java [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+      case ENTRIES:
+      case ALL_ENTRIES:
+        StructType dataFileType = (StructType) df.schema().apply("data_file").dataType();
+        DataType actualEntriesType = dataFileType.apply("partition").dataType();
+        Assert.assertEquals("Partition type must match", expectedType, actualEntriesType);
+        break;
+
+      default:
+        throw new UnsupportedOperationException("Unsupported metadata table type: " + tableType);
+
```
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply spark3/src/test/java/org/apache/iceberg/spark/source/TestMetadataTablesWithPartitionEvolution.java]: Fixed.

File: spark3/src/test/java/org/apache/iceberg/spark/source/TestMetadataTablesWithPartitionEvolution.java (hunk 6 above)
Comment (aokolnychyi) [reply spark3/src/test/java/org/apache/iceberg/spark/source/TestMetadataTablesWithPartitionEvolution.java]: Fixed. Changed to `assertPartitions`.

**Comment by aokolnychyi** [comment]
@RussellSpitzer, somehow tests for `all_data_files` started to fail after #2877. It looks like we did not update that metadata table in the PR. Is that on purpose?

Checking what is actually going on.

**Comment by aokolnychyi** [comment]
@RussellSpitzer, I think simply using `schema()` instead of `fileSchema` in `AllDataFilesTable` fixes the problem.

**Comment by RussellSpitzer** [comment]
@aokolnychyi Sounds good to me, I thought I mimicd the way we were doing it in the FilesTable using the "fileSchema" as the projected schema, if that isn't the scan.schema we should change it

**Comment by aokolnychyi** [comment]
Well, I am not sure. I'll need your help to verify whether my assumption is correct. We are using the same `ManifestReadTask` in both but `AllDataFilesTable` still passes `fileSchema` instead of `schema()`.

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 8]
```diff
@@ -177,4 +184,42 @@ public Void alwaysNull(int fieldId, String sourceName, int sourceId) {
...
+
+    for (PartitionSpec spec : table.specs().values()) {
+      List<NestedField> fields = spec.partitionType().fields();
+
+      for (NestedField field : fields) {
+        int fieldId = field.fieldId();
+        NestedField existingField = commonFields.get(fieldId);
+
+        // partition fields may conflict in v1 tables
+        ValidationException.check(existingField == null || existingField.equals(field),
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:212]: I think that this should check `fieldId`, `souceId`, and `transform`, but not `name` because it should be possible to rename fields without breaking anything. That's why we have `fieldId`.

Also, if one of the transforms is `void` then it should also be allowed because we replace transforms with a `void` transform to delete partition fields in v1. `void` produces a single value that is valid for the output of any partition function.

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 9]
```diff
@@ -177,4 +184,42 @@ public Void alwaysNull(int fieldId, String sourceName, int sourceId) {
...
+            "Conflicting partition fields: ['%s', '%s']",
+            field, existingField);
+
+        commonFields.put(fieldId, field);
+      }
+    }
+
+    List<NestedField> sortedCommonFields = commonFields.values().stream()
+        .sorted(Comparator.comparingInt(NestedField::fieldId))
+        .collect(Collectors.toList());
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:222]: +1 for sorting by `fieldId`.

**Review by kbendick — APPROVED** [review]
Overall, this looks good to me.

Same question that Ryan has regarding checking `name` during validation in `partitionType`, but overall this looks good to me. Thanks Anton!

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 10]
```diff
@@ -177,4 +186,52 @@ public Void alwaysNull(int fieldId, String sourceName, int sourceId) {
...
+  public static StructType partitionType(Table table) {
+    if (table.specs().size() == 1) {
+      return table.spec().partitionType();
+    }
+
+    Map<Integer, PartitionField> fieldMap = Maps.newHashMap();
+    List<NestedField> structFields = Lists.newArrayList();
+
+    // sort the spec IDs in descending order to pick up the most recent field names
+    List<Integer> specIds = table.specs().keySet().stream()
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:208]: Added a sort by spec ID to make sure we pick up the most recent field name (see a dedicated test too).

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+    StructType expectedType = StructType.of(
+        NestedField.optional(1000, "p2", Types.StringType.get()),
+        NestedField.optional(1001, "category", Types.StringType.get())
+    );
+    StructType actualType = Partitioning.partitionType(table);
+    Assert.assertEquals("Types must match", expectedType, actualType);
+  }
+
+  @Test
+  public void testPartitionTypeWithAddingBackSamePartitionFieldInV1Table() {
```
Comment (aokolnychyi) [inline core/src/test/java/org/apache/iceberg/TestPartitioning.java:129]: This test validates we ignore field names when building the common type. The original spec will have `1000:data` and the last spec will have `1000:data_1000` as the old field was renamed to avoid naming conflicts.

File: api/src/main/java/org/apache/iceberg/PartitionField.java [hunk 12]
```diff
@@ -67,6 +68,20 @@ public String name() {
     return transform;
   }
 
+  /**
+   * Returns true if this partition field is compatible with another partition field.
+   * <p>
+   * Partition fields are considered compatible if they have the same source ID, field ID and their
+   * transforms are equivalent or one of them is always producing nulls.
+   */
+  boolean compatibleWith(PartitionField other) {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/PartitionField.java:77]: `PartitionSpec.compatibleWith` checks that the field names match, but ignores the field ID. That allows us to dedup specs that only differ by field IDs and avoid adding a spec that isn't needed.

This checks field IDs but not field names, the opposite of the other `compatibleWith`. It isn't a good idea to have two different notions of compatibility within partitioning. I think the solution is probably to be more specific, like `equivalentIgnoringNames` or something. Not a great name for a method, but clear about what is happening. That, or we could avoid adding a method here.

File: core/src/main/java/org/apache/iceberg/Partitioning.java [hunk 13]
```diff
@@ -177,4 +186,52 @@ public Void alwaysNull(int fieldId, String sourceName, int sourceId) {
...
+        int fieldId = field.fieldId();
+        PartitionField existingField = fieldMap.get(fieldId);
+
+        if (existingField == null) {
+          fieldMap.put(fieldId, field);
+          NestedField structField = spec.partitionType().field(fieldId);
+          structFields.add(structField);
+        } else {
+          // verify the fields are compatible as they may conflict in v1 tables
+          ValidationException.check(field.compatibleWith(existingField),
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/Partitioning.java:225]: I would probably just make `equivalentIgnoringName` a private method in this class for this.

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+        .addField("data")
+        .commit();
+
+    // in v1, we use void transforms instead of dropping partition fields
+    StructType expectedType = StructType.of(
+        NestedField.optional(1000, "data_1000", Types.StringType.get()),
+        NestedField.optional(1001, "data", Types.StringType.get())
+    );
+    StructType actualType = Partitioning.partitionType(table);
+    Assert.assertEquals("Types must match", expectedType, actualType);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestPartitioning.java:149]: I think you could argue that adding `data` back should re-use the old ID. Not something to fix here, but we should probably fix it at some point.

**Review by rdblue — CHANGES_REQUESTED** [review]
Looks good to me, other than the `compatibleWith` method that conflicts with the one in `PartitionSpec`.

File: core/src/test/java/org/apache/iceberg/TestPartitioning.java (hunk 14 above)
Comment (aokolnychyi) [reply core/src/test/java/org/apache/iceberg/TestPartitioning.java:149]: It is indeed wierd. However, I would not worry too much here as we have already stated not to rename and drop fields in v1 tables.

**Comment by aokolnychyi** [comment]
Yeah, I wasn't sure about the method name. Updated.

File: api/src/main/java/org/apache/iceberg/PartitionField.java (hunk 12 above)
Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/PartitionField.java]: Fixed.

**Comment by aokolnychyi** [comment]
Thanks for reviewing, @flyrain @karuppayya @RussellSpitzer @rdblue @kbendick @jackye1995!
